### PR TITLE
www/nginx: 1.32

### PR DESCRIPTION
**Pull request by kulikov-a**
Hi!
a few suggestions have accumulated )

* add support for resetting timed out connections and 444 responses (reset_timedout_connection)
* add option to change autoban response code to 444 (NGX_HTTP_CLOSE)
* add ssl_reject_handshake directive support
* add error log severity level support for HTTP and Stream servers
* add logging of possible errors causes to the setup script
* $internalModelUseSafeDelete enabled in API Settings controller to check item references before delete (may be related to https://github.com/opnsense/plugins/issues/3196)
* minor style adjustments for IP ACL and SNI Based Routing forms
* fix: add the PROXY protocol for the HTTPS listener too, if it is set for the server (partially closes https://github.com/opnsense/plugins/issues/3159)
* fix: set Stream server outbound PROXY protocol based on Upstream settings (partially closes https://github.com/opnsense/plugins/issues/3159)
* fix: set Trusted Proxies (set_real_ip_from) for Stream server only with PROXY protocol enabled (partially closes https://github.com/opnsense/plugins/issues/3159)
  **WARNING**: set_real_ip_from directive is temporary disabled due to ngx_stream_realip module missing
* fix: do not include commented out core rules in naxsi policies when importing (ref. https://forum.opnsense.org/index.php?topic=30547.0)
* fix: fixed a typo in setting the proxy_ssl_session_reuse directive value (thanks to Sigurd Våg Aaknes) (closes https://github.com/opnsense/plugins/issues/3161)

Thanks!

### CODE REVIEW

File: www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/forms/httpserver.xml [hunk 1]
```diff
@@ -20,6 +20,13 @@
     <label>Default Server</label>
     <type>checkbox</type>
   </field>
+  <field>
+    <id>httpserver.ssl_reject_handshake</id>
+    <label>Reject SSL Handshake</label>
+    <type>checkbox</type>
+    <help>If enabled, SSL handshakes for this server will be rejected.</help>
```
Comment (fabianfrz) [inline www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/forms/httpserver.xml:27]: I never use SSL in the UI - it is just wrong.

```suggestion
    <help>If enabled, TLS handshakes for this server will be rejected.</help>
```

File: www/nginx/src/opnsense/mvc/app/models/OPNsense/Nginx/Nginx.xml [hunk 2]
```diff
@@ -52,6 +56,15 @@
         <Required>N</Required>
         <MinimumValue>1</MinimumValue>
       </server_names_hash_max_size>
+      <ban_response type="OptionField">
+        <multiple>N</multiple>
+        <OptionValues>
+          <http_403 value="403">403 Forbidden</http_403>
+          <http_444 value="444">444 Close</http_444>
```
Comment (fabianfrz) [inline www/nginx/src/opnsense/mvc/app/models/OPNsense/Nginx/Nginx.xml:63]: Maybe this way it it easier to understand:
```suggestion
          <http_444 value="444">444 Terminate Connection</http_444>
```

File: www/nginx/src/opnsense/mvc/app/models/OPNsense/Nginx/Nginx.xml [hunk 3]
```diff
@@ -698,6 +711,10 @@
           </check001>
         </Constraints>
       </default_server>
+      <ssl_reject_handshake type="BooleanField">
```
Comment (fabianfrz) [inline www/nginx/src/opnsense/mvc/app/models/OPNsense/Nginx/Nginx.xml:714]: the convention in the plugin is that the fields are called tls_* as well to have the tls also in the API calls. so please rename.
```suggestion
      <tls_reject_handshake type="BooleanField">
```

File: www/nginx/src/opnsense/mvc/app/models/OPNsense/Nginx/Nginx.xml [hunk 4]
```diff
@@ -698,6 +711,10 @@
...
+        <default>0</default>
+        <Required>Y</Required>
+      </ssl_reject_handshake>
```
Comment (fabianfrz) [inline www/nginx/src/opnsense/mvc/app/models/OPNsense/Nginx/Nginx.xml:717]:
```suggestion
      </tls_reject_handshake>
```

File: www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/forms/httpserver.xml [hunk 5]
```diff
@@ -20,6 +20,13 @@
     <label>Default Server</label>
     <type>checkbox</type>
   </field>
+  <field>
+    <id>httpserver.ssl_reject_handshake</id>
```
Comment (fabianfrz) [inline www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/forms/httpserver.xml:24]:
```suggestion
    <id>httpserver.tls_reject_handshake</id>
```

File: www/nginx/src/opnsense/scripts/nginx/setup.php [hunk 6]
```diff
@@ -75,6 +75,7 @@ function find_ca($refid)
 @chgrp('/var/db/nginx/auth', GROUP_OWNER);
 @chgrp('/var/log/nginx', GROUP_OWNER);
 $nginx = $config['OPNsense']['Nginx'];
+openlog("configd.py", LOG_ODELAY, LOG_USER);
```
Comment (fabianfrz) [inline www/nginx/src/opnsense/scripts/nginx/setup.php:78]: @AdSchellevis I guess we should not use the prefix of configd here. The script is run on init (rc.d) and by configd action. Maybe use `nginx` as prefix

File: www/nginx/src/opnsense/service/templates/OPNsense/Nginx/http.conf [hunk 7]
```diff
@@ -105,36 +108,42 @@ server {
...
 
-{%   if server.listen_https_address is defined and server.listen_https_address != '' and server.certificate is defined %}
+{%   if server.listen_https_address is defined and server.listen_https_address != '' %}
 {%     for listen_address in server.listen_https_address.split(',') %}
-    listen {{ listen_address }} http2 ssl{% if server.default_server is defined and server.default_server == '1' %} default_server{% endif %};
+    listen {{ listen_address }} http2 ssl{% if server.proxy_protocol is defined and server.proxy_protocol == '1' %} proxy_protocol{% endif %}{% if server.default_server is defined and server.default_server == '1' %} default_server{% endif %};
 {%     endfor %}
-{%     if server.ca is defined %}
+
+{%     if server.ssl_reject_handshake is defined and server.ssl_reject_handshake == '1'%}
```
Comment (fabianfrz) [inline www/nginx/src/opnsense/service/templates/OPNsense/Nginx/http.conf:116]:
```suggestion
{%     if server.tls_reject_handshake is defined and server.tls_reject_handshake == '1'%}
```

File: www/nginx/src/opnsense/service/templates/OPNsense/Nginx/streams.conf [hunk 8]
```diff
@@ -100,13 +100,14 @@
...
+        proxy_protocol {% if upstream.proxy_protocol == '1' %}on{% else %}off{% endif %};
 {%   if server.proxy_responses is defined and server.proxy_responses != '' %}
         proxy_responses {{ server.proxy_responses }};
 {%   endif%}
-{%   if server.trusted_proxies is defined and server.trusted_proxies != '' %}
+{%   if server.trusted_proxies is defined and server.trusted_proxies != '' and server.proxy_protocol is defined and server.proxy_protocol == '1' %}
 {%     for trusted_proxy in server.trusted_proxies.split(',') %}
-        set_real_ip_from {{ trusted_proxy }};
+        # temporarily disabled so as not to create a broken config with a "directive is not allowed" error. need to build with the "with-stream_realip_module" option
+        #set_real_ip_from {{ trusted_proxy }};
```
Comment (fabianfrz) [inline www/nginx/src/opnsense/service/templates/OPNsense/Nginx/streams.conf:110]: According to https://github.com/opnsense/plugins/pull/3205/files#diff-7fad4db26aa3e57d29d4f7f025a266f134a0a5bd52311263c4eab61c8f64e739R24

@fichtner and @mimugmail  are usually aware how the FreeBSD community works and maintain some packages there. If you need help or they can do it, it would be great as it would also improve the usabillty for non-OPNsense systems too.

Needs to be upstreamed to FreeBSD in the ports tree - OPNsense's fork is here, but this should go to the original:
https://github.com/opnsense/ports/blob/master/www/nginx/Makefile#L181

And then the option needs to be added here:
https://github.com/opnsense/tools/blob/master/config/23.1/make.conf#L107

**Review by fabianfrz — CHANGES_REQUESTED** [review]
looks almost good - some style changes are needed and I guess the stream thing needs a little upstream change.

File: www/nginx/src/opnsense/mvc/app/models/OPNsense/Nginx/Nginx.xml (hunk 2 above)
Comment (kulikov-a) [reply www/nginx/src/opnsense/mvc/app/models/OPNsense/Nginx/Nginx.xml:63]: sorry, mimiced nginx docs

File: www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/forms/httpserver.xml (hunk 1 above)
Comment (kulikov-a) [reply www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/forms/httpserver.xml]: sorry, mimicked nginx docs

File: www/nginx/src/opnsense/scripts/nginx/setup.php (hunk 6 above)
Comment (kulikov-a) [reply www/nginx/src/opnsense/scripts/nginx/setup.php:78]: @fabianfrz I am trying to log to the backend log here. `configd.py` is the filter on syslog-ng settings for the messages destined for this log

File: www/nginx/src/opnsense/service/templates/OPNsense/Nginx/streams.conf (hunk 8 above)
Comment (kulikov-a) [reply www/nginx/src/opnsense/service/templates/OPNsense/Nginx/streams.conf:110]: ticket is in progress for FreeBSD port https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=267112

Comment (kulikov-a) [reply www/nginx/src/opnsense/service/templates/OPNsense/Nginx/streams.conf:110]: @fichtner and @mimugmail , on the advice of @fabianfrz, I kindly ask for your advice here ). 
what is the correct way in this case: try to add the option (`--with-stream_realip_module`) to the opnsense ports repo, wait for the bugzilla ticket to close?
Thanks!

File: www/nginx/src/opnsense/scripts/nginx/setup.php (hunk 6 above)
Comment (kulikov-a) [reply www/nginx/src/opnsense/scripts/nginx/setup.php:78]: @fabianfrz @AdSchellevis 
Hi! what if i try to use syslog for Global Error Log at
https://github.com/opnsense/plugins/blob/cc42b5287871edb4dabd8384841ef3ad44705d4f/www/nginx/src/opnsense/service/templates/OPNsense/Nginx/nginx.conf#L15
repace with (`error_log  syslog:server=unix:/var/run/log;`)
and send log messages to it from setup.php too (replace tag to 'nginx').
would that be ok?

Comment (kulikov-a) [reply www/nginx/src/opnsense/scripts/nginx/setup.php:78]: @fabianfrz @AdSchellevis 
Hi! what if i try to use syslog for Global Error Log at
https://github.com/opnsense/plugins/blob/cc42b5287871edb4dabd8384841ef3ad44705d4f/www/nginx/src/opnsense/service/templates/OPNsense/Nginx/nginx.conf#L15
repace with (`error_log  syslog:server=unix:/var/run/log;`)
and send log messages to it from setup.php too (replace tag to 'nginx').
would that be ok?

Comment (AdSchellevis) [reply www/nginx/src/opnsense/scripts/nginx/setup.php:78]: @kulikov-a my preference would always be to use syslog, but I'm not involved in the nginx plugin

Comment (kulikov-a) [reply www/nginx/src/opnsense/scripts/nginx/setup.php:78]: @AdSchellevis Thanks! Your opinion matters anyway ) I'll try and see if it works correctly

Comment (kulikov-a) [reply www/nginx/src/opnsense/scripts/nginx/setup.php]: @AdSchellevis @fabianfrz how about https://github.com/opnsense/plugins/pull/3205/commits/844b859609a6b9d83c33a05525587491ddfef37e then?
Unfortunately, simply using the configuration option to use syslog has the same disadvantages as in the unbound case: syslog logging only starts to work after a successful start. and thus we lose debug startup error messages (https://github.com/opnsense/core/pull/5243). so I had to parse the file.
will that go?

Comment (AdSchellevis) [reply www/nginx/src/opnsense/scripts/nginx/setup.php]: @kulikov-a  I don't really like the error log magic, but I'm also not sure why you need it to be honest. Can't nginx send the output to syslog, or only in flagged as debug?

Comment (kulikov-a) [reply www/nginx/src/opnsense/scripts/nginx/setup.php]: @AdSchellevis maybe I didn't fully explain. nginx **can** send messages to the syslog by itself. but it starts to do this only after the master process has **successfully** loaded (that is, there should be no errors in the config file, the sockets is free etc.). if an error occurs at the start, then the diagnostic message (like "Invalid variable name ..") does not get into the log (hard to understand the reason for the failure). in the case of file logging, all startup errors gets into the log.
(this behavior resembles unbound).

Without file-to-syslog-magic, separate diagnostic scripts will be required (at least config file testing imho).
or may be we can just leave the initial option (keep logging to file and send setup.php diagnostic messages to the backend log ("configd.py" prefix))

Comment (AdSchellevis) [reply www/nginx/src/opnsense/scripts/nginx/setup.php]: @kulikov-a I understand your problem, but in case of unbound it's just a feed to `logger` which is quite clean, in this case it has hard to follow magic, it might be cleaner to feed information back in the startup/apply via the gui as it's most likely an issue was caused there. 

But as I mentioned earlier, I'm not involved in the Nginx plugin (nor do I want to be), any solution likely works for me.

Comment (kulikov-a) [reply www/nginx/src/opnsense/scripts/nginx/setup.php]: @AdSchellevis sorry for bother. I just decided that if it affects the core componentsl, then it is better to get your opinion. if you're not happy with this syslog "magic", then I'll look into other options. and if you and @fabianfrz  don't mind just using the `configd.py` tag in setup script, then I would stop there for now (I'll try to add at least a "config test" to the gui before migrating to syslog).
testing small changes in the processhandler, which should allow to get the stderr of configd commands in the backend log (https://github.com/kulikov-a/core/commit/884c8bedf91b7ff0bf4ed7e6e3e51d02f91e0c84) and . seems to work

Comment (AdSchellevis) [reply www/nginx/src/opnsense/scripts/nginx/setup.php]: @kulikov-a help? I think @fabianfrz suggestion about the target was better, technically this isn't a `configd` message. I kind of got distracted by the rest of the fuzz in the ticket. I guess at some point in time I do need to spend some time on `configd` itself as well, still have some wishes on my list, but priority hasn't been high enough yet.  As usual, if additions are small and easy to verify (and add something useful), we might merge it, sometimes it just takes time.

Comment (kulikov-a) [reply www/nginx/src/opnsense/scripts/nginx/setup.php]: @AdSchellevis hi! yes, thanks! sorry again - i distracted you and I myself seem to be a little confused (although tbh i was glad when I came up with the parser ) ). I will update the pr tomorrow with the standard use of syslog and add a config check with syslog feeding in rc.d conf. this will at least add information about config errors to the log (other startup errors will hopefully end up in the backend log).
Thanks!

File: www/nginx/src/opnsense/service/templates/OPNsense/Nginx/streams.conf (hunk 8 above)
Comment (fabianfrz) [reply www/nginx/src/opnsense/service/templates/OPNsense/Nginx/streams.conf:110]: If you would ask me I would prefer to implement that ticket without unimplemented placeholders so I would prefer to wait until the patch is upstream.

Comment (kulikov-a) [reply www/nginx/src/opnsense/service/templates/OPNsense/Nginx/streams.conf:110]: thanks, sorry if I didn’t fully understand: since the `streamserver.trusted_proxies` field is already in the model and template (and I just tried to use it more correctly), probably in this case it makes sense to temporarily hide it from the form (until the `stream_realip_module` module appears in the assembly), so as not to confuse the users?
did I understand correctly?
(hopefully I'll have a chance to try out the patch-Makefile soon and discuss the possibility of using it with @fichtner )

**Comment by kulikov-a** [comment]
@fabianfrz Hi!
stream_realip_module added to FreeBSD port and already synced by @fichtner at https://github.com/opnsense/ports/commit/cf3a52bcaeb738e75b7b11475ff24c849a006d7d
PR  has been updated to reflect the changes. plus added the mentioned but strangely forgotten update of the naxsi rule import script.
please take a look when you have time.
Thanks!

**Review by fabianfrz — APPROVED** [review]
looks good - you could have put the shell part into the php setup script but that's a style question.

File: www/nginx/src/opnsense/service/templates/OPNsense/Nginx/nginx [hunk 9]
```diff
@@ -1,2 +1,2 @@
-nginx_setup="/usr/local/opnsense/scripts/nginx/setup.php"
+nginx_setup="/usr/local/opnsense/scripts/nginx/setup.php; nginx -t -q 2>&1 | logger -t nginx -p user.emerg"
```
Comment (fichtner) [inline www/nginx/src/opnsense/service/templates/OPNsense/Nginx/nginx:1]: I agree with @fabianfrz -- we can move this addition to setup.php with a shell_safe() call

**Review by fichtner — APPROVED** [review]
expected the whole command pipe in shell_safe() as a one line change but this works as well 👍

**Comment by fichtner** [comment]
Merged, thanks!

**Comment by kulikov-a** [comment]
thanks everyone!
